shiny-api/common/kafka.go

```diff
@@ -0,0 +1,299 @@
+package common
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"log"
+	"math/rand"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/pkg/errors"
+)
+
+const (
+	CreateCommandType string = "create"
+	UpdateCommandType string = "update"
+)
+
+type RequestPayload struct {
+	ID        string    `json:"id,omitempty"`
+	Name      string    `json:"name,omitempty"`
+	Foo       float64   `json:"foo,omitempty"`
+	CreatedOn time.Time `json:"created_on,omitempty"`
+	UpdatedOn time.Time `json:"updated_on,omitempty"`
+	Version   string    `json:"version,omitempty"`
+}
+
+type SuccessPayload struct {
+	ID        string    `json:"id"`
+	Name      string    `json:"name"`
+	Foo       float64   `json:"foo"`
+	CreatedOn time.Time `json:"created_on"`
+	UpdatedOn time.Time `json:"updated_on"`
+	Version   string    `json:"version"`
+}
+
+type ErrorPayload struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+}
+
+type ThingCommandEntry struct {
+	CID     string          `json:"cid"`
+	Type    string          `json:"type"`
+	Request *RequestPayload `json:"request"`
+	Success *SuccessPayload `json:"success"`
+	Failure *ErrorPayload   `json:"failure"`
+
+	encoded []byte
+	err     error
+}
+
+func (te *ThingCommandEntry) ensureEncoded() {
+	if te.encoded == nil && te.err == nil {
+		te.encoded, te.err = json.Marshal(te)
+	}
+}
+
+func (te *ThingCommandEntry) Length() int {
+	te.ensureEncoded()
+	return len(te.encoded)
+}
+
+func (te *ThingCommandEntry) Encode() ([]byte, error) {
+	te.ensureEncoded()
+	return te.encoded, te.err
+}
+
+func (te *ThingCommandEntry) String() string {
+	if te == nil {
+		return "nil"
+	}
+
+	return fmt.Sprintf(
+		"cid:'%s' type:%s request:%+v success:%+v failure:%+v",
+		te.CID,
+		te.Type,
+		te.Request,
+		te.Success,
+		te.Failure,
+	)
+}
+
+type KafkaClient struct {
```

Comment:
Generally speaking, in golang methods are only preferred when you are messing with internal state as a side effect (or satisfying an interface).  Given that you aren't really doing much of that, it probably makes more sense to move to functions that take as arguments the things they need to do their behavior.